types/connector.d.ts

```diff
@@ -3,28 +3,384 @@
+   * @param options Discovery options
+   * @param cb Callback function
+   */
+  discoverForeignKeys?(modelName: string[], options: IDPropertiesDiscoveryOptions, cb: Callback<DiscoveredForeignKeys>): Promise<DiscoveredForeignKeys>;
+  /**
+   * {@inheritDoc Connector.discoverForeignKeys}
+   * @deprecated
+   */
+  discoverForeignKeysSync?(modelName: string[], options: IDPropertiesDiscoveryOptions): DiscoveredForeignKeys;
+
+  /**
+   * Retrieve a description of the foreign key columns that reference the given
+   * table's primary key columns (i.e. The foreign keys exported by a table),
+   * ordered by `fkOwner`, `fkTableName`, and `keySeq`.
+   * 
+   * @param modelName Target model name
+   * @param options Discovery options
+   * @param cb Callback function
+   */
+  discoverExportedForeignKeys?(modelName: string, options: IDPropertiesDiscoveryOptions, cb: Callback<DiscoveredForeignKeys>): Promise<DiscoveredForeignKeys>;
+
+  /**
+   * {@inheritDoc Connector/discoverExportedForeignKeys}
+   * @deprecated
+   */
+  discoverExportedForeignKeysSync?(modelName: string, options?: {owner?: string}): DiscoveredForeignKeys;
+
+  /**
+   * Discover schema from a given table name / view name.
+   * 
+   * @param tableName Target table name
+   * @param options Discovery options
+   * @param cb Callback function
+   */
+  discoverSchemas(tableName: string, options: SchemaDiscoveryOptions, cb: Callback<Schema>): Promise<Schema>;
+
+  /**
+   * Check whether or not migrations are required for the database schema to match
+   * the model definitions attached to the {@link DataSource}.
+   * 
+   * @param models Name of models to check. If not defined, all models are checked.
+   * @param cb 
+   */
+  isActual?(models?: string | string[], cb?: Callback<boolean>): void;
+
+  /**
+   * Freeze the datasource. Behaviour depends on the {@link Connector}.
+   */
+  freezeDataSource?(): void;
+
+  /**
+   * {@inheritDoc Connector.freezeDataSource}
+   * 
+   * @remarks
+   * This is kept for backwards-compatibility with JugglingDB connectors.
+   * Connectors should implement {@link Connector.freezeDataSource}.
+   */
+  freezeSchema?(): void;
+
+  /**
+   * Normalize connector-specific return data into standardised Juggler context
+   * data.
+   * 
+   * @remarks
+   * Depending on the connector, the database response can contain information
+   * about the updated record(s). This object usually has a database-specific
+   * structure and does not match model properties. For example, MySQL returns
+   * `OkPacket: {fieldCount, affectedRows, insertId, ... , changedRows}`.
+   * 
+   * The return value is normalised data.
+   * 
+   * If the connector DDL and DML functions map directly to a hash-map of
+   * model properties and their values, this function does not need to be
+   * implemented.
+   * 
+   * @param context 
+   * @param dbResponse 
+   */
+  generateContextData?(context: Context, dbResponse: unknown): Context;
+
```

Comment:
Add missing migration functions:

dropTable(model, cb)
alterTable(model, actualFields, actualIndexes, cb)
createTable(model, cb)
addIndexes(model, actualIndexes)
showIndexes(model, cb)
showFields(model, cb)
getColumnsToAdd(model, actualFields)
getColumnsToDrop(model, actualFields)
getAddModifyColumns(model actualFields)
propertiesSQL(model) <-- Alias in PostgreSQL: buildColumnDefinitions

The above were found in MSSQL and PostgreSQL connectors.